crates/core/tedge_mapper/src/mapper.rs

```diff
@@ -85,14 +89,121 @@ impl Mapper {
             let _ = self.output.send(init_message).await;
         }
 
-        while let Some(message) = &mut self.input.next().await {
-            let converted_messages = self.converter.convert(&message);
-            for converted_message in converted_messages.into_iter() {
-                let _ = self.output.send(converted_message).await;
-            }
+        // Detect and reconcile any alarms that this mapper may have missed while not running
+        self.reconcile_missed_alarms().await;
```

Comment:
Beware that all the mapper processes are running this code (c8y, az and collectd).

So I would delegate this to the converter. I agree that's not easy: the converter can have a state but should not be given a direct access to mqtt.

Reply:
This would be the heart of the new design and here's how I'm planning to do this. I'll have to introduce 3 states that the converter can be in:
1. Caching phase: Where all alarms received on the startup , from the tedge/alarms and the internal/alarms topics, are cached into their respective maps without doing an immediate conversion. Messages received on other topics will be simply converted, as they would be during the normal phase as well.
2. Reconciliation phase: Both the maps produced during the caching phase are compared and the differences are converted and sent to the respective cloud topics.
3. The normal phase: This is the regular phase during which the converter would just be doing the simple stateless message mapping.

The mapper main would be the one driving the converter through these stages based on the timer (the reconciliation window). The converter would have no notion of time. It just sees itself in different states and reacts accordingly.

Reply:
Your proposal sounds good.

My remark was more about the fact that this `reconcile_missed_alarms` loop is called by *all* the mappers and that *only* the c8y mapper should reconcile alarms for c8y. With the current code base, the collectd mapper will also reconcile the alarms ... its own way, meaning that the internal alarms will be updated but nothing forwarded.